analyzer/block/block.go

```diff
@@ -0,0 +1,275 @@
+// common logic for queueing blocks and support for parallel processing.
+package block
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+
+	"github.com/oasisprotocol/oasis-indexer/analyzer"
+	"github.com/oasisprotocol/oasis-indexer/analyzer/queries"
+	"github.com/oasisprotocol/oasis-indexer/analyzer/util"
+	"github.com/oasisprotocol/oasis-indexer/config"
+	"github.com/oasisprotocol/oasis-indexer/log"
+	"github.com/oasisprotocol/oasis-indexer/storage"
+)
+
+const (
+	// Timeout to process a block.
+	processBlockTimeout = 61 * time.Second
+	// Number of blocks to be processed in a batch.
+	blocksBatchSize = 100
+	// Lock expire timeout for blocks (in minutes). Locked blocks not processed within
+	// this time can be picked again.
+	lockExpiryMinutes = 5
+)
+
+// BlockProcessor is the interface that block-based processors should implement to use them with the
+// block based analyzer.
+type BlockProcessor interface {
+	// PreWork performs tasks that need to be done before the main processing loop starts.
+	PreWork(ctx context.Context) error
+	// ProcessBlock processes the provided block, retrieving all required information
+	// from source storage and committing an atomically-executed batch of queries
+	// to target storage.
+	//
+	// The implementation must commit processed blocks (update the chain.processed_blocks record with processed_time timestamp).
+	ProcessBlock(ctx context.Context, height uint64) error
+	// SourceLatestBlockHeight returns the latest block height available in the source storage.
+	SourceLatestBlockHeight(ctx context.Context) (uint64, error)
+}
+
+var _ analyzer.Analyzer = (*blockBasedAnalyzer)(nil)
+
+type blockBasedAnalyzer struct {
+	config       *config.BlockBasedAnalyzerConfig
+	analyzerName string
+
+	processor BlockProcessor
+
+	target storage.TargetStorage
+	logger *log.Logger
+}
+
+// firstUnprocessedBlock returns the first block before which all blocks have been processed.
+func (b *blockBasedAnalyzer) firstUnprocessedBlock(ctx context.Context) (first uint64, err error) {
+	err = b.target.QueryRow(
+		ctx,
+		queries.FirstUnprocessedBlock,
+		b.analyzerName,
+	).Scan(&first)
+	return
+}
+
+// unlockBlock unlocks a block.
+func (b *blockBasedAnalyzer) unlockBlock(ctx context.Context, height uint64) {
+	rows, err := b.target.Query(
+		ctx,
+		queries.UnlockBlockForProcessing,
+		b.analyzerName,
+		height,
+	)
+	if err != nil {
+		rows.Close()
+	}
+}
+
+// fetchBatchForProcessing fetches (and locks) a batch of blocks for processing.
+func (b *blockBasedAnalyzer) fetchBatchForProcessing(ctx context.Context, from, to uint64) ([]uint64, error) {
+	// XXX: In future, use a system for picking lock IDs in case other parts of the code start using advisory locks.
+	const lockID = 1001
+	var (
+		tx      storage.Tx
+		heights []uint64
+		rows    pgx.Rows
+		err     error
+	)
+
+	// Start a transaction.
+	tx, err = b.target.Begin(ctx)
```

Comment:
We could (should?) skip the explicit transaction; I don't think it achieves anything here, as we're doing a single query besides the (un)locking.

Per [docs](https://www.postgresql.org/docs/current/explicit-locking.html#ADVISORY-LOCKS): "Unlike standard lock requests, session-level advisory lock requests do not honor transaction semantics: a lock acquired during a transaction that is later rolled back will still be held following the rollback, and likewise an unlock is effective even if the calling transaction fails later."

Probably better to move the unlocking into `defer`? It's not foolproof but postgres does drop the lock at the end of the session.

**Edit:**: I found `LOCK TABLE` ([docs](https://www.postgresql.org/docs/current/sql-lock.html)) which is per-transaction and more closely matches what we need, compared to advisory locks. So let's keep the transaction and use that instead?

Reply:
Actually there's also 'pg_advisory_xact_lock' - transaction level advisory locks. Can probably just use that.

^ I thought transaction level locks are automatically used if used within a transaction, but as you pointed out that's not the case.

Reply:
Neat, I didn't see `pg_advisory_xact_lock`. I think let's use that.
For every mega-query, there will be `batch_size` writes to the table (to update `processed_time` to `now()`), and it will indeed be better if those are not locked out.
I'm fine with either though; we can always change later if we notice problems.